Add release notes for 0.285 (#21579)
prestodb/presto: presto-docs/src/main/sphinx/release/release-0.285.rst

=============
Release 0.285
=============

**Details**
===========

Security Changes
________________

> * Fix critical vulnerability in Babel and related npm packages by updating to newer versions. :pr:`21322`

Author: @tdcmeehan I felt it is easier to read the release notes if we keep them brief and include a link to the PR for more details.
Any thoughts?

Reviewer: I think this is a good idea to keep the note concise, while allowing people to dig deeper.

=============
Release 0.285
=============

**Details**
===========

General Changes
_______________
* Fix `min_by(x, y, n) / max_by(x, y, n)` functions to ensure `n` is unique.
* Fix LIKE pattern to handle multi-byte characters. :pr:`21578`
* Improve latency of HBO. :pr:`21297`
* Improve HBO to track partial aggregation statistics. :pr:`21160`
* Improve PushPartialAggregationThroughExchange to use partial aggregation statistics when available. This extends to multi-key aggregations.
* Improve union node by removing the empty inputs.
* Improve HBO latency by skipping retry if it times out. :pr:`21308`

> * Improve join-reordering to work with non-simple equi-join predicates.

Reviewer: ```suggestion
* Improve cost based optimizer join reordering to work with ....
```

Any suggestions on how to make "non-simple equijoin predicates" a little bit more understandable?

Author: Probably a link to the PR will help.

Author: CC: @aaneja for input.

Reviewer 2: How about - `Improve join reordering for complex equi-join clauses` PR : https://github.com/prestodb/presto/pull/21153

Author: `complex` is slightly better than `non-simple` but I believe a Presto user is better off looking at the PR for the actual improvement. I don't think there is an easy wording here.

> * Improve :doc:`REST API</rest>` by providing links.

Reviewer 3: ```suggestion
* Improve :doc:`REST API</rest>` documentation by providing links.
```
Suggest adding the word "documentation" to the three doc entries - with the loss of the heading the context for these, when grouped into General Changes, feels missing. It's not immediately obvious when reading the page that these three entries are docs-only changes and didn't affect code.

Author: Good catch!

> * Improve :doc:`REST API</rest>` documentation by providing links.

Reviewer: ```suggestion
```
Just a docs update, can omit.

> * Improve :doc:`Resource Groups </admin/resource-groups>` documentation.

Reviewer: ```suggestion
```
Docs

> * Add Presto to MySQL type mapping in :doc:`MySQL Connector</connector/mysql>`.

Reviewer 3: ```suggestion
* Add Presto to MySQL type mapping in :doc:`MySQL Connector</connector/mysql>` documentation.
```

> * Add Presto to MySQL type mapping in :doc:`MySQL Connector</connector/mysql>` documentation.

Reviewer: ```suggestion
```

* Add :func:`array_least_frequent` function.
* Add support for HANA connector. :pr:`21034`
* Add task killer which is triggered when a worker is running out of memory and the garbage collector cannot reclaim sufficient memory. Two strategies are provided: full garbage collection, and frequent full garbage collection. :pr:`21254`
* Add a session property ``history_input_table_statistics_matching_threshold`` to set the threshold for input statistics match in HBO. :pr:`21063`
* Add a session property ``remove_redundant_cast_to_varchar_in_join`` (enabled by default) to remove redundant `cast to varchar` expressions in a join condition. :pr:`21050`
* Add a session property ``enable_hbo_for_scaled_writer`` (disabled by default) to use HBO for scaled writers.

> * Add a session property ``use_partial_aggregation_history`` if partial aggregation statistics must be used to split aggregates into partial and final. :pr:`21160`

Reviewer: Let's word it like:

Feature -> this feature is configurable by session property X (enabled/disabled by default)

> * Add a config property ``task.memory-based-slowdown-threshold`` that provides a heap memory threshold below which new splits are not processed. :pr:`20946`

Reviewer: This feature is experimental and dangerous, I do not recommend adding it to the release notes.

> * Add tracking of null probe keys in join and use it to enable outer join null salt in history-based optimization. :pr:`21203`

Reviewer: This needs to be more understandable for a user.
```suggestion
* Improve history based optimizer performance when nulls are present in joins. :pr:`21203`
```

* Add optimization where values node followed by an always false filter is converted to an empty values node.
* Add information about cost-based optimizers and the source of stats they use (CBO/HBO) in explain plans when session property ``verbose_optimizer_info_enabled=true``. :pr:`20990`

> * Add a boolean to the PlanOptimizer's return value to indicate if an optimization has been triggered.

Reviewer: This is not a user facing change, can remove

> * Rename config ``experiemental.task.high-memory-task-killer-strategy`` to ``experimental.task.high-memory-task-killer-strategy``.

Reviewer: Per above, let's just remove it altogether

* Upgrade Joda-Time to 2.12.5 to use 2023c tzdata. The JVM should also be updated to ensure the timezone data is consistent. :pr:`21329`
* Upgrade Alluxio to 304 from 2.8.1.
* Upgrade AWS SDK to 1.12.560 from 1.12.261.
* Upgrade Avro version to 1.11.3 from 1.9.2.

Prestissimo (Native Execution) Changes
______________________________________
* Fix task cleanup to use task termination time instead of task end time.
* Add support for JWT authentication. :pr:`20290`
* Add session property ``native_debug.validate_output_from_operators`` to identify malformed output from operators. :pr:`21036`

Security Changes
________________
* Fix critical vulnerability in Babel and related npm packages by updating to newer versions. :pr:`21322`

Hive Connector Changes
______________________
* Improve support for schema evolution of partition column types for Parquet file format. :pr:`19983`

> * Add support for Parquet writer versions V1 and V2. See :doc:`Hive Connector</connector/hive>`.

Reviewer 3: Thought about suggesting changing this link to 
`Hive Connector <../connector/hive.html#parquet-writer-version>`_  to link directly to the heading on the page, but  because the top of the Hive Connector doc page has "Parquet Writer Version" in the table of contents list of headings, I don't think it really adds value: the link you found and added is great. 

No change suggested, just unpacking my thought process.

* Add a boolean to the PlanOptimizer's return value to indicate if an optimization has been triggered.
* Improve latency of HBO. :pr:`21297`
* Improve HBO to track partial aggregation statistics. :pr:`21160`
* Improve PushPartialAggregationThroughExchange to use partial aggregation statistics when available. This extends to multi-key aggregations.
* Improve union node by removing the empty inputs.
* Skip HBO retry when it timeouts.
* Improve join-reordering to work with non-simple equi-join predicates.
* Fix `min_by(x, y, n) / max_by(x, y, n)` functions to ensure `n` is unique.
* Track null probe keys in join and use it to enable outer join null salt in history based optimization.
* Update Joda-Time to 2.12.5 to use 2023c tzdata. The JVM should also be updated to ensure the timezone data is consistent. See :pr:`21329` for more details.
* Update Alluxio to 304 from 2.8.1.
* Update AWS SDK to 1.12.560 from 1.12.261.
* Update Avro version to 1.11.3 from 1.9.2.

Hive Connector Changes
______________________
* Improve support for schema evolution of partition column types for Parquet file format. :pr:`19983`

> * Add support for Parquet writer versions V1 and V2 in Hive and Iceberg catalogs.

Reviewer 3: ```suggestion
* Add support for Parquet writer versions V1 and V2 in Hive and Iceberg catalogs. :pr:`20957`
```
I'd link to the doc but it doesn't seem to exist yet per [this comment on PR 21500](https://github.com/prestodb/presto/pull/21500#discussion_r1431890364). Maybe it would be appropriate to create a new issue for adding this requested doc.

Author: It was merged here https://github.com/prestodb/presto/pull/21587

Reviewer 3: Thanks for finding that and adding the link to the doc!
